src/tags.ts

```diff
@@ -0,0 +1,83 @@
+/**
+ * @file User tags.
+ * @author Johan Nordberg <johan@steemit.com>
+ */
+
+import {JsonRpcAuthMethodContext as JCtx, JsonRpcError} from '@steemit/koa-jsonrpc'
+import * as config from 'config'
+import {Op} from 'sequelize'
+
+import {Tag, UserTag} from './database'
+
+const ADMIN_ACCOUNT = config.get('admin_role') as string
+
+export async function defineTag(this: JCtx, name: string, description: string) {
+    this.assert(this.account === ADMIN_ACCOUNT, 'Unauthorized')
+    this.assert(typeof name === 'string', 'Invalid tag name')
+    this.assert(typeof description === 'string' && description.length > 0, 'Invalid tag description')
+    await Tag.create({name, description})
+}
+
+export async function listTags(this: JCtx) {
+    this.assert(this.account === ADMIN_ACCOUNT, 'Unauthorized')
+    return (await Tag.all({order: ['name']})).map(({name, description}) => ({name, description}))
+}
+
+export async function assignTag(this: JCtx, uid: string, tag: string, memo: string) {
+    this.assert(this.account === ADMIN_ACCOUNT, 'Unauthorized')
+    this.assert(typeof uid === 'string', 'Invalid user uid')
+    this.assert(typeof tag === 'string', 'Invalid tag')
+    if (typeof memo !== 'string') {
+        memo = `Created by ${ this.account } from ${ this.ctx.ip }`
+    }
+    const existing = await UserTag.find({where: {uid, tag, deletedAt: {[Op.eq]: null}}})
+    if (!existing) {
+        try {
+            await UserTag.create({uid, tag, memo})
+        } catch (error) {
+            if (error.name === 'SequelizeForeignKeyConstraintError') {
+                throw new JsonRpcError(420, 'No such tag')
+            }
+            throw error
+        }
+    }
```

Comment:
I considered this but landed on user already having the tag not being an error since that's the result you're looking for by calling the function, e.g. redeemer assigning tags would have to implement specific error checking for TagAlreadyAssigned errors otherwise. If you can think of some use case where this would be a problem I'm open to changing it.